lib/containers/basetest.pm

```diff
@@ -0,0 +1,43 @@
+# SUSE's openQA tests
+#
+# Copyright 2021 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Base class for container tests
+# Maintainer: qac team <qa-c@suse.de>
+
+package containers::basetest;
+use base 'opensusebasetest';
+use testapi;
+use strict;
+use warnings;
+use containers::common qw(install_docker_when_needed install_podman_when_needed);
+use version_utils 'get_os_release';
+
+sub containers_factory {
+    my ($self, $runtime) = @_;
+    my $engine;
+
+    die("Egnine already initialized") if ($self->{engine});
+
+    my ($running_version, $sp, $host_distri) = get_os_release;
+
+    if ($runtime eq 'docker') {
+        $engine = containers::engine::docker->new();
+        install_docker_when_needed($host_distri);
+    }
+    elsif ($runtime eq 'podman') {
+        $engine = containers::engine::podman->new();
+        install_podman_when_needed($host_distri);
```

Comment:
not completely convince about that. i sort like the idea but i think it is not that simple. you mix the design with other logic. But the design should stay as clean and simply as possible to take the maximum benefits. i am afraid to propose something as it is not one or two that can be the alternatives. 
In brief keep in mind: the design pattern should focus and solve the class relationships, when the inserted subroutine focus on the configuration of the system, so i would say it is another level. flexibility is nice but i believe we need another way to show that.

Reply:
I can leave this in the test module as it was, and we can think about this later on. 
My idea is also to move this into engine.pm and do something like `$engine->install()` or `$engine->setup()`.

Reply:
See [poo#101444](https://progress.opensuse.org/issues/101444) for that